test/upgrade/README.md

```diff
@@ -0,0 +1,83 @@
+
+This is inspired by kubernetes
+[upgrade testing](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/e2e-tests.md#version-skewed-and-upgrade-testing).
+
+These tests are a pretty big hammer in that they cover more than just version
+changes, but it’s one of the only ways to make sure we don’t accidentally make
+breaking changes for now.
+
+## Flow
+
+We’d like to validate that the upgrade doesn’t break any resources (they still
+propagate events) and doesn't break our installation (we can still update
+resources).
+
+At a high level, we want to do this:
+
+1. Install the latest knative release.
+1. Create some resources.
+1. Install knative at HEAD.
+1. Test those resources, verify that we didn’t break anything.
+
+To achieve that, we just have three separate build tags:
+
+1. Install the latest release from GitHub.
+1. Run the `preupgrade` tests in this directory.
+1. Install at HEAD (`ko apply -f config/`).
+1. Run the `postupgrade` tests in this directory.
+1. Install the latest release from GitHub.
+1. Run the `postdowngrade` tests in this directory.
+
+## Tests
+
+### Smoke test
+
+This was stolen from the e2e tests as one of the simplest cases.
+
+#### preupgrade, postupgrade, postdowngrade
+
+Run the selected smoke test.
+
+### Probe test
+
+In order to verify that we don't have data-plane unavailability during our
+control-plane outages (when we're upgrading the knative/eventing installation),
+we run a prober test that continually sends events to a service during the
+entire upgrade/downgrade process. When the upgrade completes, we make sure that
+all of those events propagated just once.
+
+To achieve that a tool was prepared (https://github.com/cardil/wathola). It
+consists of 3 components: _sender_, _forwarder_, and _receiver_. _Sender_ is the
+usual Kubernetes pod that publishes events to the default `broker` with given
+interval. When it closes (by either `SIGTERM`, or `SIGINT`), an `finished` event
+is generated. _Forwarder_ is a knative serving service that scales from zero to
+receive the requested traffic. _Forwarders_ receive events and forwards them to
+given target. _Receiver_ is an ordinary pod that collects events from multiple
+forwarders and has an endpoint `/report` that can be polled to get the status of
+sent events.
+
+Diagram below describe the setup:
+
+```
+  (pod)              (ksvc)             (pod)
++--------+       +-----------+       +----------+
+|        |       |           ++      |          |
+| Sender |   +-->| Forwarder ||----->+ Receiver |
+|        |   |   |           ||      |          |
++---+----+   |   +------------|      +----------+
+    |        |    +-----------+
+    |        |
+    |        |
+    |     +--+-----+       +---------+
+    +----->        |       |         +-+
+          | Broker | < - - | Trigger | |
+          |        |       |         | |
+          +--------+       +---------+ |
+           (default)        +----------+
+```
```

Comment:
This documentation is great, I'd like to see a section on what the output is, where to look for debug hints, look like you dump a bunch of logs, etc. For example you report that there are double events, my first knee jerk reaction is, are there for example double-subscriptions being created somehow. Not quite sure how I'd go about figuring it out. Please create an issue to follow up so we don't have to hold / rebase, etc.

Reply:
In the e2e tests we dump the objects / k8s events on test failures so something like that would be nice. But again, fine as a followup.

Reply:
Now, we are not failing even that we found errors. If we switch https://github.com/cardil/eventing/blob/feature/upgrade-tests/test/upgrade/probe_test.go#L55 then we will have default fail dump after this test. Maybe that is sufficient?

Reply:
That line L55 gives me 404.

Reply:
Yup. I've removed the source branch, sorry. But, now it's merged it's on master: https://github.com/knative/eventing/blob/master/test/upgrade/probe_test.go#L55